Client Config Builder Module (Refactored)

rihp opened the pull request:
Continued from #53, enables the Client Config Builder Module following the requests from @cbrzn  
- https://github.com/polywrap/python-client/pull/53#discussion_r1029360329

The new implementation uses the `ClientConfig` as a dataclass to later create an Abstract Base Class of a `BaseClientConfigBuilder` which defines more clearly the functions of the module, like add_envs, set_resolvers, remove_wrappers, and so on.  

This `BaseClientConfigBuilder` is later used in the class `ClientConfigBuilder` which only implements the build method, for now, and inherits all the abstract methods of the `BaseClientConfigBuilder`.

- Closes: #5
- Related issues: #96 

## Features
- [x] build
- [x] add
- [x] get envs
- [x] Set_env
- [x] add_env
- [x] add_envs
- [x] add interface implementations
- [x] set interface implementations
- [x] set package
- [x] add package
- [x] add packages
- [x] remove packages
- [x] add wrapper
- [x] add wrappers
- [x] remove wrapper
- [x] set_resolver
- [x] set_redirect
- [x] set_redirects
- [x] remove_redirect
- [x] Readme

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,185 @@
+    This Abstract class is used to configure the polywrap client before it executes a call
+    The ClientConfig class is created and modified with the ClientConfigBuilder module
+    """
+    envs: Dict[Uri, Dict[str, Any]]
+    interfaces: Dict[Uri, List[Uri]]
+    resolver: List[IUriResolver]
+    wrappers: List[Uri]
```
```
export interface ClientConfig<TUri extends Uri | string = Uri | string> {
    readonly envs: Env<TUri>[];
    readonly interfaces: InterfaceImplementations<TUri>[];
    readonly wrappers: IUriWrapper<TUri>[];
    readonly packages: IUriPackage<TUri>[];
    readonly resolvers: UriResolverLike[];
}
```
Fix this according to js ccb

Niraj-Kamdar replied:
Without redirects though

cbrzn replied:
i think it make sense to have redirects in builder configuration

cbrzn replied:
fwiw js client config still has redirects https://github.com/polywrap/toolchain/blob/origin-dev/packages/js/client-config-builder/src/ClientConfig.ts#L14

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,185 @@
+        pass
+
+    @abstractmethod
+    def get_envs(self)-> Dict[Uri, Dict[str, Any]]:
```
If it has implementation, it isn't an abstractmethod

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,185 @@
+    @abstractmethod    
+    def set_env(self, env: Env, uri: Uri):
+        """Sets the envs dictionary in the builder's config, overiding any existing values."""
+        if (env or uri) is None:
```
it can't be None if it's not Optional

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,185 @@
+        return self
+
+    @abstractmethod
+    def add_env(self, env: Env = None , uri: Uri = None):
```
It shouldn't be default to None if we don't want it to be Optional

rihp replied:
updated

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,185 @@
+        """
+        return self.config
+    
+    def get_envs(self) -> Dict[Uri, Dict[str, Any]]:
```
You don't need to implement inherited methods unless you want to change it

Niraj-Kamdar reviewed (changes requested):
The implementation isn't according to the latest js changes. Please update it to be in parity with js client

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,256 @@
+    envs: Dict[Uri, Dict[str, Any]]
+    interfaces: Dict[Uri, List[Uri]]
+    packages: List[UriPackage]
+    resolver: List[IUriResolver]
```
That should be UriResolverLike

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,256 @@
+    def __init__(self):
+        self.config = ClientConfig(envs={}, interfaces={}, resolver=[], wrappers= [], packages=[])
+
+    @abstractmethod
```
only build is the abstractmethod

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,256 @@
+        pass
+
+    @abstractmethod
+    def add(self, new_config: ClientConfig) -> ClientConfig:
```
if you have implementation than it's not an abstractmethod

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,256 @@
+    def add_env(self, env: Env, uri: Uri):
+        """Adds an environment (in the form of an `Env`) for a given uri, without overwriting existing environments,
+        unless the env key already exists in the environment, then it will overwrite the existing value"""
+        if (env or uri) is None:
```
this is unnecessary

Niraj-Kamdar commented on packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py:
```diff
@@ -0,0 +1,256 @@
+    """
+    envs: Dict[Uri, Dict[str, Any]]
+    interfaces: Dict[Uri, List[Uri]]
+    packages: List[UriPackage]
```
I believe this can also be optimised to a dict instead of list of object. Can be done in the separate issue though

rihp commented:
I've pushed progress again, adding more through testing, cleanups, fixtures, removing abstractclass decorators from functions with an implementation, refactored the client_config.packages property, added new tests for it, and also a first e2e test of the generic ccb.add(partial_client_config)
https://github.com/polywrap/python-client/pull/54

rihp commented:
Applied lint changes and new requests from Cesar

Niraj-Kamdar reviewed (changes requested):
Checkout https://github.com/polywrap/toolchain/tree/origin-dev/packages/js/client-config-builder since it's been changed a lot after this PR.  
Some notable changes:
- BuilderConfig
- Only CoreClientConfig remained
- build signature changed

Niraj-Kamdar reviewed (approved):
Thanks @rihp for the contributions! I have disabled the test for now but will be re-enabling when I tackle exhaustive test tasks since it requires changes due to refactors!
